File: src/DynamicLinkLibraries/Motion6D/Motion6D.Portable/VisibleConsumerLink.cs
using CategoryTheory;
using Diagram.UI.Interfaces;
using Motion6D.Interfaces;
using NamedTree;
using System;

namespace Motion6D.Portable
{
    /// <summary>
    /// Link between visible object and its consumer
    /// </summary>
    public class VisibleConsumerLink : ICategoryArrow, IDisposable, IPostSerialize, IAllowCodeCreation
    {

        #region Fields

        /// <summary>
        /// Associated object
        /// </summary>
        protected object obj;

        /// <summary>
        /// Consumer
        /// </summary>
        protected IVisibleConsumer source;
 

        /// <summary>
        /// Visible object
        /// </summary>
        protected IVisible target;

 

        #endregion

        #region Ctor

        /// <summary>
        /// Default constructor
        /// </summary>
        public VisibleConsumerLink()
        {
        }

        #endregion

        #region ICategoryArrow Members

        ICategoryObject ICategoryArrow.Source
        {
            get
            {
                return source as ICategoryObject;
            }
            set
            {
                if (!(value is IPosition))
                {
                    CategoryException.ThrowIllegalSourceException();
                }
                IPosition p = value as IPosition;
                if (value is IVisibleConsumer)
                {
                    source = value as IVisibleConsumer;
                    return;
                }
                if (p.Parameters == null)
                {
                    CategoryException.ThrowIllegalSourceException();
                }
                object o = p.Parameters;
                if (!(o is IVisibleConsumer))
                {
                    CategoryException.ThrowIllegalSourceException();
                }
                source = o as IVisibleConsumer;
            }
        }

        ICategoryObject ICategoryArrow.Target
        {
            get
            {
                return target as ICategoryObject;
            }
            set
            {
                if (value is IPosition)
                {
                    IPosition p = value as IPosition;
                    if (p.Parameters == null)
                    {
                        CategoryException.ThrowIllegalSourceException();
                    }
                    object o = p.Parameters;
                    if (o is IVisible)
                    {
                        target = o as IVisible;
                        source.Add(target);
                    }
                }
                else
                {
                    CategoryException.ThrowIllegalTargetException();
                }
            }
        }

        #endregion

        #region IAssociatedObject Members

        object IAssociatedObject.Object
        {
            get
            {
                return obj;
            }
            set
            {
                obj = value;
            }
        }

        bool IAllowCodeCreation.AllowCodeCreation => false;

        #endregion

        #region IDisposale Members

        /// <summary>
        /// Removing of VisibleConsumerLink
        /// </summary>
        void IDisposable.Dispose()
        {
            source.Remove(target);
        }

        #endregion

        #region IPostSerialize Members

        void IPostSerialize.PostSerialize()
        {
            source.Post(target);
        }

        #endregion


   /*!!!FICTION
        void Fiction()
        {
            IVisible visible = null;
            IVisibleConsumer consumer = null;
            ICategoryArrow link = new VisibleConsumerLink();
            link.Source = consumer as ICategoryObject;
            link.Target = visible as ICategoryObject;

            consumer.Add(visible);
            
        }
    */


    }
}

File: src/TypeScript/TypeScriptBackup/src/Library/Motion6D/Visible/VisibleConsumerLink.ts
import { CategoryArrow } from "../../CategoryArrow";
import type { ICategoryObject } from "../../Interfaces/ICategoryObject";
import type { IDesktop } from "../../Interfaces/IDesktop";
import type { IVisible } from "./Interfaces/IVisible";
import type { IVisibleConsumer } from "./Interfaces/IVisibleConsumer";

export class VisibleConsumerLink extends CategoryArrow {
    constructor(desktop: IDesktop, name: string) {
        super(desktop, name);
        this.typeName = "VisibleConsumerLink";
        this.types.push("VisibleConsumerLink");
    }

    getSource(): ICategoryObject {
        return this.consumer as unknown as ICategoryObject;
    }

    getTagret(): ICategoryObject {
        return this.visible as unknown as ICategoryObject;
    }

    setSource(source: ICategoryObject): void {
        var c = this.performer.convertProperties<IVisibleConsumer>(source, "IVisibleConsumer")
        this.consumer = c[0]
    }

    setTarget(target: ICategoryObject): void {
        this.visible = this.performer.convertProperties<IVisible>(target, "IVisible")[0]
        this.consumer.addVisibleObject(this.visible)
       }

    consumer!: IVisibleConsumer;

    visible!: IVisible;

}